torchao/prototype/inductor/fx_passes/int8_sdpa_fusion.py

```diff
@@ -0,0 +1,581 @@
+            v_scale,
+            a_zp,
+            a_scale,
+            o_zp,
+            o_scale,
+        )
+        trans_output = L[aten.permute.default](output, [0, 2, 1, 3])
+        return L[aten.clone.default](
+            trans_output, memory_format=torch.contiguous_format
+        )
+
+    return int8_sdpa
+
+
+def _get_int8_sdpa_q_pattern(is_batch_size_1: bool, has_convert: bool):
+    int8_sdpa_q_basic_pattern = CallFunction(
+        torch.ops.quantized_decomposed.dequantize_per_tensor.default,
+        CallFunction(
+            aten.permute.default,
+            KeywordArg("query"),
+            Arg(),
+        ),
+        KeywordArg("q_scale"),
+        KeywordArg("q_zp"),
+        Arg(),
+        Arg(),
+        Arg(),
+    )
+    if has_convert:
+        int8_sdpa_q_basic_pattern = CallFunction(
+            torch.ops.prims.convert_element_type.default,
+            int8_sdpa_q_basic_pattern,
+            Arg(),
+        )
+    int8_sdpa_q_basic_pattern = CallFunction(
+        aten.expand.default,
+        int8_sdpa_q_basic_pattern,
+        Arg(),
+    )
+    if is_batch_size_1:
+        # pattern is different for bs=1
+        return CallFunction(
+            aten.reshape.default,
+            int8_sdpa_q_basic_pattern,
+            Arg(),
+        )
+    else:
+        return CallFunction(
+            aten.reshape.default,
+            CallFunction(
+                aten.clone.default,
+                int8_sdpa_q_basic_pattern,
+                memory_format=Arg(),
+            ),
+            Arg(),
+        )
+
+
+def _get_int8_sdpa_k_pattern(is_batch_size_1: bool, has_convert: bool):
+    int8_sdpa_k_basic_pattern = CallFunction(
+        torch.ops.quantized_decomposed.dequantize_per_tensor.default,
+        CallFunction(
+            aten.permute.default,
+            CallFunction(
+                aten.permute.default,
+                KeywordArg("key"),
+                Arg(),
+            ),
+            Arg(),
+        ),
+        KeywordArg("k_scale"),
+        KeywordArg("k_zp"),
+        Arg(),
+        Arg(),
+        Arg(),
+    )
+    if has_convert:
+        int8_sdpa_k_basic_pattern = CallFunction(
+            torch.ops.prims.convert_element_type.default,
+            int8_sdpa_k_basic_pattern,
+            Arg(),
+        )
+    int8_sdpa_k_basic_pattern = CallFunction(
+        aten.expand.default,
+        int8_sdpa_k_basic_pattern,
+        Arg(),
+    )
+    if is_batch_size_1:
+        # pattern is different for bs=1
+        return CallFunction(
+            aten.reshape.default,
+            int8_sdpa_k_basic_pattern,
+            Arg(),
+        )
+    else:
+        return CallFunction(
+            aten.reshape.default,
+            CallFunction(
+                aten.clone.default,
+                int8_sdpa_k_basic_pattern,
+                memory_format=Arg(),
+            ),
+            Arg(),
+        )
+
+
+def _get_int8_sdpa_v_pattern(is_batch_size_1: bool, has_convert: bool):
```

Comment:
Can we merge the `q/k/v` pattern into one function?

Reply:
Thanks, modified!